### DISCUSSION TITLE
Wails instead of Electron

### DISCUSSION THREAD

**Discussion by vaaski** [wavetermdev/waveterm#1523] [Ideas]
Hey, I just stumbled across this repo and looking at the languages used, I thought I was looking at a wails project.

Turns out this is based on electron instead, but I think wails would be a perfect fit.
I am not associated with the wails project but I love the idea of ditching electron without having to learn rust for tauri.

So this is just a suggestion to [check wails out](https://github.com/wailsapp/wails) :)

**Comment by esimkowitz** [comment]
Funny you mention it, we actually did a pretty deep investigation into Wails when we were rearchitecting our app this summer, however we didn't find it sufficient for production use, unfortunately.

I'm copying over some old Discord messages summarizing of our findings and rationale for sticking with Electron:

## [From @sawka](https://discord.com/channels/977384027629690900/1208513848676720731/1270227698232459265) 

appreciate the feedback.  we actually did a deep dive into using Wails (same technology as Tauri, using webviews, except using a Go backend).  surprisingly the performance was much worse than electron.  it caught us by surprise, and we tried multiple configurations (sidecars, in-proc, out-of-proc, etc.).  the only advantage (performance wise) was that it had very small download sizes.  i think tauri/wails can be good for some apps, but performance and the robustness of the ecosystem is much better on electron.  cross-platform story is much better as well.

there's a lot of talk about security as well.  most of that talk is completely outdated and only applies if you run electron in an insecure mode (exposing nodejs bindings into the renderer processes -- which of course we dont do).  running electron using isolation gives us the same protections that you get running chrome (or Tauri).  Also, the fact that Google owns Chrome, and there is Slack (Salesforce), Discord, VSCode (Microsoft), etc. all running on electron means it gets security updates fast.  I have no reason to believe Tauri is insecure, but I know that Chromium has a lot more eyeballs on it and a lot more vested interest in keeping it patched and up to date.

## [From me](https://discord.com/channels/977384027629690900/1208513848676720731/1270514911742529627) 

Also the impacts of the download/install size for Electron are minimized on Linux when using a package manager, since the underlying Electron binaries get reused. (We're still working on getting the package manager entries set up, but those should be coming soon!)

For macOS, a big part of the large size is the universal binary, which is in fact not universal, it just packages the arm64 and amd64 binaries together in one distributable. This means you are basically downloading two apps (and two Chromium binaries, and two Node binaries), which sucks. We're working on separating out the two architectures, but first we need to update our website to detect your local architecture and serve the correct download link. 

I was pushing really hard for us to try and get Wails working, but yeah as Mike said it wasn't even close (we're talking backend calls that would take seconds on Wails and milliseconds on Electron, animations that dropped more frames than they rendered, etc.). We will absolutely keep an eye on these projects to see if they manage to address some of our concerns. If it ever reaches a maturity level that satisfies us, we'll do our best to migrate over. I'd love to be in a place where we can contribute to that success, but we're a really small team so for now we can't afford to make such investments when there's a really performant option already available.

**Reply by vaaski** [reply]
Interesting stuff, makes sense. Thanks!

**Reply by esimkowitz** [reply]
I was really bummed out about it, I really wanted Wails to work. I'm still hopeful it can in the future, but the way they handle IPC between the frontend and the backend is just not good enough for us. The one that was probably the most egregious was the way they'd process backend calls. We'd see multiple calls to the backend would come in and we'd respond to each of them in like 2ms, but then the Wails would take like another 500ms to process the response and return it to the caller and they'd all come in at once, even if the calls were made out of order. Just weird stuff that I hope will get sorted out as they mature.

**Reply by vaaski** [reply]
I would've never expected the non-JS solution to be slower than electron. Is the wails team aware of this?

**Reply by esimkowitz** [reply]
Y'know I don't think we ended up reporting it to them :/ we were heads-down and it just slipped my mind.

**Reply by esimkowitz** [reply]
The thing with Electron (and Node/V8 in general) is that it's actually crazy fast, it's just not very efficient. Chrome has 1000s of engineers working on optimizing the littlest things to eke out that much more perf from it (especially for networking calls, which we do a lot of) and it pays off for a remarkably stable engineering target. If all you're doing is rendering a ToDo app or something like 1Password (which uses Tauri) and you don't need to do a ton of CRUD or streaming, then those trade-offs are probably fine to make. For us, though, where we need maximum throughput and graphics more than we need efficiency, it wasn't worth it.
